BUG: `lfiltic`'s handling of `a[0] != 1` differs from `lfilter`'s

lciti opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
Closes #22470.

#### What does this implement/fix?
Fixes 22470 by dividing the resulting `zi` by `a[0]`.
Also added checks that `a` is non-empty and its first element is non-zero.
Added tests to test the issue before and after the fix (also added additional test cases for `lfiltic`).

#### Additional information
<!--Any additional information you think is important.-->
None

Commit:
Added test to highlight issue #22470 prior to fixing it

Commit:
BUG: lfiltic's handling of a[0] != 1 differs from lfilter's

Fixes 22470 by dividing the resultin zi by a[0].
Also added checks that a is non-empty and its first element is non-zero

Closes #22470

j-bowhay commented on scipy/signal/_signaltools.py:
```diff
@@ -2367,6 +2367,9 @@ def lfiltic(b, a, y, x=None):
     y = xp.asarray(y)
 
+    if N < 0:
+        raise ValueError("There must be at least one `a` coefficient.")
```
Please can you add test coverage for this

j-bowhay commented on scipy/signal/_signaltools.py:
```diff
@@ -2391,14 +2394,20 @@ def lfiltic(b, a, y, x=None):
+    a0 = a[0] if N > 0 else a
+    if a0 != 1.:
+        if a0 == 0.:
+            raise ValueError("First `a` filter coefficient must be non-zero.")
```
And this

Commit:
Addressed CI error and added test coverage for ValueErrors (as requested by j-bowhay)

lciti commented:
@ev-br I am getting some errors in `lfiltic` (see [here](https://github.com/scipy/scipy/actions/runs/13415830954/job/37476370519#step:13:3604)) around the "concat = array_namespace(a).concat" line. I was wondering if you could help me understand why that line is needed in the first place. Can't we just use "x = **xp.**concat((x, xp.zeros(M - L)))" as it's done a couple of lines below for "y"?

ev-br commented:
> I am getting some errors in lfiltic (see here) around the "concat = array_namespace(a).concat" line. I was wondering if you could help me understand why that line is needed in the first place.

This pattern is needed when `xp` is not fully array API compatible. For instance, `numpy.concat` is new in numpy 2.0 So to run with numpy == 1.x, you go through `array_api_compat.numpy` which wraps numpy proper and adds Array API compatible functions. 

You can add these kinds of wrappers in a reactive way: if the CI is OK on numpy 2.0 and fails on 1.x, you just add this stanza to where it fails. 

The errors on CI though are different though:

```
 a          = array([1.0, 0.5, 1.5], dtype=object)
```

is wrong. Object arrays should not be created, so your best bet is to trace where this comes from and fix it to be a normal numerical array.

Commit:
Small changes to address CI errors

lciti commented:
@ev-br Thank you for your reply. The "object" type is due to the way the test is already set up (it tries different data types, including objects) and has nothing to do with my edit. I have set that line to "xp.concat" for now since this is similar to a change made to "y" after your initial commit. That test seems to work now.

Commit:
Fixed failing test with cupy and excluded test with jax.
The problem with jax is that does not support inplace ops.
It could be easily fixed, e.g. with zi = zi.at[idx].add(...)
Another option would be to rewrite parts of the code to use
(xp or scipy) convolution instead.

Commit:
Fixed Lint error UP038 Use `X | Y` in `isinstance` call instead of `(X, Y)`

lciti commented:
@j-bowhay I have taken your comments onboard and I have done a few minor changes to address the CI issues with non-numpy array APIs. It seems to pass all the tests now.

lciti commented:
Hi, just a friendly ping on this PR. I've addressed all comments and would appreciate it if you could let me know if there's anything else needed or if it's ready to be merged. Thanks!

j-bowhay commented:
Thanks for making the changes, it would be good if one of the signal regulars could take a look

larsoner reviewed (approved):
I am not knowledgeable about the `xp`-related changes but the signal changes look good to me!

j-bowhay reviewed (approved):
Looks good from an array api perspective and there is approval from a signal regular, thanks @lciti, keep them coming!

tylerjereddy commented:
I considered for backporting but it seemed too messy according to my note in gh-22840, so I'll remove the label.

Ping me if there is a concern about that, but I need to stop dragging my feet on the `1.15.3` release so getting organized for that.
